compiler/noirc_frontend/src/elaborator/enums.rs

```diff
@@ -243,4 +252,670 @@ impl Elaborator<'_> {
+
+            self.unify(&body_type, &result_type, || TypeCheckError::TypeMismatch {
+                expected_typ: result_type.to_string(),
+                expr_typ: body_type.to_string(),
+                expr_span: body_span,
+            });
+
+            self.pop_scope();
+            Row::new(columns, guard, body)
+        });
+        (rows, result_type)
+    }
+
+    /// Convert an expression into a Pattern, defining any variables within.
+    fn expression_to_pattern(&mut self, expression: Expression, expected_type: &Type) -> Pattern {
+        let expr_span = expression.span;
+        let unify_with_expected_type = |this: &mut Self, actual| {
+            this.unify(actual, expected_type, || TypeCheckError::TypeMismatch {
+                expected_typ: expected_type.to_string(),
+                expr_typ: actual.to_string(),
+                expr_span,
+            });
+        };
+
+        match expression.kind {
+            ExpressionKind::Literal(Literal::Integer(x, negative)) => {
+                let actual = self.interner.next_type_variable_with_kind(Kind::IntegerOrField);
+                unify_with_expected_type(self, &actual);
+                Pattern::Int(SignedField::new(x, negative))
+            }
+            ExpressionKind::Literal(Literal::Bool(value)) => {
+                unify_with_expected_type(self, &Type::Bool);
+                let constructor = if value { Constructor::True } else { Constructor::False };
+                Pattern::Constructor(constructor, Vec::new())
+            }
+            ExpressionKind::Variable(path) => {
+                // A variable can be free or bound if it refers to an enum constant:
+                // - in `(a, b)`, both variables may be free and should be defined, or
+                //   may refer to an enum variant named `a` or `b` in scope.
+                // - Possible diagnostics improvement: warn if `a` is defined as a variable
+                //   when there is a matching enum variant with name `Foo::a` which can
+                //   be imported. The user likely intended to reference the enum variant.
+                let path_len = path.segments.len();
+                let location = Location::new(path.span(), self.file);
+                let last_ident = path.last_ident();
+
+                match self.resolve_path_or_error(path) {
+                    Ok(resolution) => self.path_resolution_to_constructor(
+                        resolution,
+                        Vec::new(),
+                        expected_type,
+                        location.span,
+                    ),
+                    Err(_) if path_len == 1 => {
+                        // Define the variable
+                        let kind = DefinitionKind::Local(None);
+                        // TODO: `allow_shadowing` is false while I'm too lazy to add a check that we
+                        // don't define the same name multiple times in one pattern.
```

Comment:
Is there an issue captured for this?

Reply:
Or maybe not need to capture anything... I see a bunch of `todo!` below. I guess the idea is to fix them in follow-up PRs, right?

Reply:
> Or maybe not need to capture anything... I see a bunch of todo! below. I guess the idea is to fix them in follow-up PRs, right?

Right. I made an issue anyway though since this was a smaller item that'd be easier to forget.